code/modules/ninja/ninja_explosive.dm

```diff
@@ -10,41 +10,72 @@
 	name = "spider charge"
 	desc = "A modified C-4 charge supplied to you by the Spider Clan.  Its explosive power has been juiced up, but only works in one specific area."
 	boom_sizes = list(4, 8, 12)
-	var/mob/detonator = null
+	///Weakref to the mob that has planted the charge
+	var/datum/weakref/detonator
+	///The only area that the charge is allowed to be planted, and detonated in
+	var/area/detonation_area
 
-/obj/item/grenade/c4/ninja/afterattack(atom/movable/AM, mob/user, flag)
-	var/datum/antagonist/ninja/ninja_antag = user.mind.has_antag_datum(/datum/antagonist/ninja)
-	if(!ninja_antag)
-		to_chat(user, span_notice("While it appears normal, you can't seem to detonate the charge."))
+/obj/item/grenade/c4/ninja/Destroy()
+	detonator = null
+	detonation_area = null
+	return ..()
+
+/**
+ * set_detonation_area
+ *
+ * Proc used to set the allowed location for charge detonation
+ *
+ * Arguments
+ * * datum/antagonist/ninja/ninja_antag - The antag datum for the owner of the c4
+ */
+/obj/item/grenade/c4/ninja/proc/set_detonation_area(datum/antagonist/ninja/ninja_antag)
+	if (!ninja_antag)
 		return
-	if (!check_loc(user, ninja_antag))
+	var/datum/objective/plant_explosive/objective = locate() in ninja_antag.objectives
+	if (!objective)
 		return
-	detonator = user
+	detonation_area = objective.detonation_location
+
+/obj/item/grenade/c4/ninja/afterattack(atom/movable/AM, mob/ninja, flag)
+	if(!IS_SPACE_NINJA(ninja))
+		to_chat(ninja, span_notice("While it appears normal, you can't seem to detonate the charge."))
+		return
+	if (!check_loc(ninja))
+		return
+	detonator = WEAKREF(ninja)
 	return ..()
 
 /obj/item/grenade/c4/ninja/detonate(mob/living/lanced_by)
-	var/datum/antagonist/ninja/ninja_antag = detonator.mind.has_antag_datum(/datum/antagonist/ninja)
-	if(!check_loc(detonator, ninja_antag)) // if its moved, deactivate the c4
-		new /obj/item/grenade/c4/ninja(target.loc)
+	if(!check_loc(detonator.resolve())) // if its moved, deactivate the c4
+		var/obj/item/grenade/c4/ninja/new_c4 = new /obj/item/grenade/c4/ninja(target.loc)
+		new_c4.detonation_area = detonation_area
+		new_c4.say("Invalid location!")
 		target.cut_overlay(plastic_overlay, TRUE)
 		qdel(src)
 		return
 	. = ..()
 	//Since we already did the checks in afterattack, the denonator must be a ninja with the bomb objective.
 	if(!detonator)
 		return
+	var/mob/ninja = detonator.resolve()
+	var/datum/antagonist/ninja/ninja_antag = ninja.mind.has_antag_datum(/datum/antagonist/ninja)
```

Comment:
`weakref.resolve()` will return null if the object is deleted.

You have to factor this in. `if (isnull(ninja)) return` will probably work fine here.